src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs

```diff
@@ -386,5 +390,39 @@ private void ResolveReferences(PEReader peReader, ImmutableHashSet<string>? refe
                 }
             }
         }
+
+        /// <inheritdoc />
+        public void LogAllDiagnostics(string? headerMessage = null)
+        {
+            if (HasRoslynDiagnostics(out IReadOnlyList<Diagnostic> roslynDiagnostics))
+            {
+                if (!string.IsNullOrEmpty(headerMessage))
+                {
+                    _logger.LogWarning(headerMessage!);
+                }
+
+                foreach (Diagnostic warning in roslynDiagnostics)
+                {
+                    _logger.LogWarning(warning.Id, warning.ToString());
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public void LogAllWarnings(string? headerMessage = null)
+        {
+            if (HasLoadWarnings(out IReadOnlyList<AssemblyLoadWarning> loadWarnings))
+            {
+                if (!string.IsNullOrEmpty(headerMessage))
+                {
+                    _logger.LogWarning(headerMessage!);
+                }
+
+                foreach (AssemblyLoadWarning warning in loadWarnings)
+                {
+                    _logger.LogWarning(warning.DiagnosticId, warning.Message);
+                }
+            }
+        }
```

Comment:
I think these would be more suitable as extension methods for the IAssemblySymbolLoader interface but I don't have a strong opinion. @ericstj what do you think?

Reply:
> Maybe we didn't have log abstraction at the time? I wonder if we can just rethink these methods based on usage and testing requirements, given the current app architecture.

Yes, we only later provided log abstractions that weren't coupled to APICompat specifics (suppressions).

Reply:
I had to look into this closer and here's a draft of what I was imagining. Basically what you mentioned above. Inline the logging into AssemblySymbolLoader and remove the unnecessary interface methods. We don't need the AssemblyLoadWarning type so I deleted it. I can't think of any _current_ need for an object model around it.

https://github.com/dotnet/sdk/commit/3d70789a824ae5ab5f0c6ccde855b498742432bc

The diagnostics one is a bit ugly as it would return the same diagnostics again if you would call the method again but we don't use any of the public methods twice in our codebase. Honestly, I haven't yet seen any use case for the diagnostics and would be fine with removing it.

Reply:
I see that you're logging warnings separately in your commit, instead of putting the warning logging code inside a method. I mention this because all this code is intended to be reused by ApiDiff as well.

Reply:
Okay, looking closer, if I am going to refactor this code, it makes sense to have these two very specfici warnings in the code that loads the assembly and looks for the references. In other words, seems it does not need to be handled _generically_, but only for these two cases.

Anything else gets handled by the diagnostic logging method you created.

Reply:
Unless someone has any objections, would you like to apply your commit to this PR?

Reply:
> seems it does not need to be handled generically, but only for these two cases.

Correct.

> Anything else gets handled by the diagnostic logging method you created.

I don't really like the diagnostics logging but you can keep it in if you feel them useful. I have never seen a roslyn diagnostic emitted by the new GenAPI during execution.

> Unless someone has any objections, would you like to apply your commit to this PR?

Of course. But you will need to update some tests. It's just a draft.

Reply:
Yeah I can update the tests.

> I don't really like the diagnostics logging but you can keep it in if you think they are useful. I have never seen a roslyn diagnostic emitted by the new GenAPI during execution.

I'm seeing some diagnostics in the ApiDiff project :( but maybe if I fix them we can remove it (later).

Reply:
Oh yeah, I can see the test failures after applying the commit. We were using those methods all over the place! 😄

Reply:
Done.